### DISCUSSION TITLE
Adding support for procedural programs to `drasil-code`

### DISCUSSION THREAD

**Discussion by B-rando1** [JacquesCarette/Drasil#3871] [General]
As mentioned in #3868, once we move the Julia renderer over from `OOProg` to `ProcProg`, we lose the ability to generate Julia code because `drasil-code` only knows how to generate `OOProg` code.  We need to enable `drasil-code` to generate procedural programs.

I don't think the 'main part' of the implementation will be too bad - a lot of things in `drasil-code` that currently require `OOProg` can be relaxed to `SharedProg`, which means that all renderers can use them.

I have two things where I'm less sure, though:
- How do we want to organize the distinction between `OOProg` and `SharedProg`?  The example I'm currently looking at is `genCode`, the root function for generating code.  Specifically this part:
https://github.com/JacquesCarette/Drasil/blob/c86e1175eb09ac8a4b53f684ba20ad71c7bc6b83/code/drasil-gen/lib/Language/Drasil/Generate.hs#L151-L158
This works in the `main` branch because `unJLC` is of the same type as `unJC` and the rest, but in #3868 and beyond it's of a different type.
  - Would it make sense to have this function, and any others in a similar situation, take an `Either`?  This would allow us to pass it `Right JC` or `Left JLC`, and let it go from there.
- How do we want to enforce the types of `Choices` that work with Julia?
  - For now, we could just throw errors 'on the fly', whenever the renderer tries to generate code that it can't.
  - It would be better to catch these sooner.  We could do this either by analyzing `Choices` before starting, or by somehow constraining `Choices` to only allow certain options for certain types of output languages.
---
Edit: I just realized that at least for this function, we don't need `Either`.  We can split up the `genCall` function into `genCallOO` and `genCallProc` to handle the input type signatures, and do the same for `generateCode`.  Since `genCall` and `generateCode` simply return `IO ()`, we don't need to worry about differences in output types.

**Comment by B-rando1** [comment]
I did some poking around in `drasil-code`.  There are a total of `99` functions that required an instance of `OOProg`.  Of those, only `18` of those constraints could be relaxed to `SharedProg`.  This isn't what I had hoped for, as it means at least `81` functions will need to be split up in order to support `ProcProg` in `drasil-code`.

I'll take a stab at splitting a few up, and see how it goes.

**Comment by JacquesCarette** [comment]
I think splitting the `genCall` function (and `generateCode`) is a better idea.

I agree that we should check for valid `Choices` earlier. But that should be a subsequent PR, this one shouldn't be blocked by that.

My gut feel is that close inspection of those `81` functions will yield opportunities for improvements that will result in more things being able to go into `SharedProg`.

**Comment by B-rando1** [comment]
I made the necessary changes to `drasil-code` to have it generating Julia code again, which you can see in #3868.  I didn't find any more functions that can 'naturally' be relaxed to `SharedProg`.  The issue is that they all use things like `mkVal`, `mkVar`, `convTypeOO`, `convExpr`, etc. that all have a chance of giving an object back.

For some of the functions that call these, we can probably assume that when they call one of these functions they _won't_ get an object back, but these assumptions aren't enforced by the underlying types.  If we want to merge more of these functions, we need to either:
- Leave the underlying types alone, and investigate each function on a case-by-case basis.  If we can be reasonably sure that in a certain case a function call won't return anything OO-related, we can switch the call to a `SharedProg` version.
- Investigate the types, and separate them into their shared, OO, and (eventually) procedural components.  This will make the types harder to organize, but will hopefully make it easier to merge functions.

**Reply by JacquesCarette** [reply]
Definitely the second option is the way to go. I'd start with the leaf-most types and work my way up. This may well be best done in a dozen PRs, some of which may feel like they barely do anything -- but each will get incrementally closer.

**Reply by B-rando1** [reply]
I've been avoiding this for a while, but I figured I should try to take it a bit further if I can.  I wasn't sure how to start, so I did another pass over the functions in question.

It looks like there are around 60 problematic functions left.  I noted the reasons why each one is problematic.  If you're interested I'll include them here.

<details>
<summary>Complete list of functions, with reasons for separation</summary>
<p><code>FunctionCalls.hs</code>:</p>
<ul>
<li><code>genAllInputCalls</code>: relies on <code>genInputCall</code>, <code>genDerivedCall</code>, and <code>genConstraintCall</code>.</li>
<li><code>genDerivedCall</code>: relies on <code>inOutCall</code></li>
<li><code>genConstraintCall</code>: relies on <code>genFuncCall</code></li>
<li><code>genCalcCall</code>: relies on <code>genFuncCall</code></li>
<li><code>genOutputCall</code>: relies on <code>genFuncCall</code></li>
<li><code>genFuncCall</code>: relies on <code>mkVal</code> and <code>fApp</code></li>
<li><code>genInOutCall</code>: relies on <code>mkVar</code> and <code>fAppInOut</code>
<code>Import.hs</code>:</li>
<li><code>value</code>: relies on <code>variable</code> and <code>convExpr</code></li>
<li><code>variable</code>:<ul>
<li>Relies on <code>inputVariable</code> and <code>constVariable</code></li>
<li>OO uses <code>staticConst</code> but Proc uses <code>constant</code></li>
</ul>
</li>
<li><code>inputVariable</code>: May return an object if inputs are bundled</li>
<li><code>constVariable</code>: May return an object if constants are bundled</li>
<li><code>mkVal</code>:<ul>
<li>Relies on <code>value</code></li>
<li>If the <code>CodeVarChunk</code> has an object, it returns an object</li>
</ul>
</li>
<li><code>mkVar</code>:<ul>
<li>Relies on <code>variable</code></li>
<li>If the <code>CodeVarChunk</code> has an object, it returns an object</li>
</ul>
</li>
<li><code>genModDef</code>:<ul>
<li>Relies on <code>genModuleWithImports</code>, <code>genFunc</code>, and <code>publicFunc</code></li>
<li>Different types of <code>Module</code></li>
</ul>
</li>
<li><code>mkParam</code>: relies on <code>mkVar</code></li>
<li><code>publicFunc</code>: relies on <code>genMethod</code></li>
<li><code>privateMethod</code>/<code>privateFunction</code>:<ul>
<li>Relies on <code>genMethod</code></li>
<li>For OO it&#39;s a method; for Proc it&#39;s a function</li>
</ul>
</li>
<li><code>genMethod</code>: relies on <code>mkVar</code> and <code>mkParam</code></li>
<li><code>genFunc</code>:<ul>
<li>Relies on <code>convStmt</code>, <code>mkVar</code>, and <code>genDataFunc</code></li>
<li>Only OO can do constructors</li>
</ul>
</li>
<li><code>genModFuncs</code>: relies on <code>genFunc</code> and <code>publicFunc</code></li>
<li><code>readData</code>: relies on <code>mkVal</code>, <code>mkVar</code>, and <code>getEntryVarLogs</code></li>
<li><code>getEntryVars</code>: relies on <code>mkVar</code>, <code>variable</code>, and <code>convType</code></li>
<li><code>getEntryVarLogs</code>: relies on <code>getEntryVars</code></li>
<li><code>convExpr</code>:<ul>
<li>Relies on <code>convCall</code>, and <code>fApp</code></li>
<li>Only OO can do constructors calls, method calls, and object field access.</li>
</ul>
</li>
<li><code>convCall</code>: relies on <code>convExpr</code> and <code>mkVar</code></li>
<li><code>convStmt</code>: relies on <code>convExpr</code> and <code>mkVar</code></li>
<li><code>genDataFunc</code>: relies on <code>readData</code> and <code>publicFunc</code></li>
<li><code>publicInOutFunc</code>: relies on <code>genInOutFunc</code></li>
<li><code>privateInOutMethod</code>/<code>privateInOutFunc</code>:<ul>
<li>Relies on <code>genInOutFunc</code></li>
<li>For OO it&#39;s a method; for Proc it&#39;s a function</li>
</ul>
</li>
<li><code>genInOutFunc</code>: relies on <code>mkVar</code>
<code>GenerateGOOL.hs</code>:</li>
<li><code>genModuleWithImports</code>: modules are different types and have different parameters</li>
<li><code>genModule</code>: relies on <code>genModuleWithImports</code></li>
<li><code>fApp</code>: Only OO does method calls</li>
<li><code>fAppInOut</code>: Only OO does method calls
<code>Modules.hs</code>:</li>
<li><code>genMain</code>:<ul>
<li>Relies on <code>genModule</code> and <code>genMainFunc</code></li>
<li>Different kinds of <code>File</code></li>
</ul>
</li>
<li><code>genMainFunc</code>: relies on <code>mkVar</code>, <code>initConsts</code>, <code>getInputDecl</code>, <code>genAllInputCalls</code>, <code>genCalcCall</code>, and <code>genOutputCall</code></li>
<li><code>initConsts</code>:<ul>
<li>Relies on <code>mkVar</code> and <code>convExpr</code></li>
<li>OO returns an object in constants are bundled</li>
</ul>
</li>
<li><code>genInputMod</code>:<ul>
<li>Relies on <code>genModule</code>, <code>genInputFormat</code>, <code>genInputDerived</code>, and <code>genInputConstraint</code></li>
<li>Different types of <code>File</code></li>
</ul>
</li>
<li><code>getInputDecl</code>:<ul>
<li>Relies on <code>mkVar</code></li>
<li>OO returns an object if inputs are bundled</li>
</ul>
</li>
<li><code>genCalcMod</code>:<ul>
<li>Relies on <code>genModuleWithImports</code> and <code>genCalcFunc</code></li>
<li>Different types of <code>File</code></li>
</ul>
</li>
<li><code>genCalcFunc</code>: relies on <code>mkVar</code>, <code>genCalcBlock</code>, <code>convStmt</code>, and <code>publicFunc</code></li>
<li><code>genCalcBlock</code>: relies on <code>genCaseBlock</code>, <code>mkVar</code>, and <code>convExpr</code></li>
<li><code>genCaseBlock</code>: relies on <code>convExpr</code> and <code>genCalcBlock</code></li>
<li><code>genInputFormat</code>: relies on <code>publicInOutFunc</code>, <code>privateInOutMethod</code>/<code>privateInOutFunc</code>, and <code>readData</code></li>
<li><code>genInputDerived</code>: relies on <code>publicInOutFunc</code>, <code>privateInOutMethod</code>/<code>privateInOutFunc</code>, and <code>genCalcBlock</code></li>
<li><code>genInputConstraints</code>: relies on <code>publicFunc</code>, <code>privateMethod</code>/<code>privateFunc</code>, <code>sfwrCBody</code>, and <code>physCBody</code></li>
<li><code>sfwrCBody</code>: relies on <code>chooseCstr</code></li>
<li><code>physCBody</code>: relies on <code>chooseCstr</code></li>
<li><code>chooseCstr</code>: relies on <code>convExpr</code>, <code>constrWarn</code>, and <code>constrExc</code></li>
<li><code>constrWarn</code>: relies on <code>constraintViolatedMsg</code></li>
<li><code>constrExc</code>: relies on <code>constraintViolatedMsg</code></li>
<li><code>constraintViolatedMsg</code>: relies on <code>printConstraint</code> and <code>mkVal</code></li>
<li><code>printConstraint</code>: relies on <code>convExpr</code></li>
<li><code>genOutputMod</code>:<ul>
<li>Relies on <code>genModule</code> and <code>genOutputFormat</code></li>
<li>Different types of <code>File</code></li>
</ul>
</li>
<li><code>genOutputFormat</code>: relies on <code>mkVal</code> and <code>publicFunc</code>
<code>Generator.hs</code>:</li>
<li><code>generateCode</code>:<ul>
<li>Relies on <code>genPackage</code></li>
<li>Different types of <code>Program</code></li>
</ul>
</li>
<li><code>genPackage</code>:<ul>
<li>Relies on <code>genProgram</code></li>
<li>Different types of <code>Program</code></li>
</ul>
</li>
<li><code>genProgram</code>:<ul>
<li>Relies on <code>chooseModules</code></li>
<li>Different types of <code>Program</code></li>
</ul>
</li>
<li><code>chooseModules</code>:<ul>
<li>Relies on <code>genUnmodular</code> and <code>genModules</code></li>
<li>Different types of <code>File</code></li>
</ul>
</li>
<li><code>genUnmodular</code>:<ul>
<li>Relies on <code>genModuleWithImports</code>, <code>genMainFunc</code>, <code>genCalcFunc</code>, <code>genModFuncs</code>, <code>genInputFormat</code>, <code>genInputDerived</code>, <code>genInputConstraints</code>, and <code>genOutputFormat</code></li>
<li>Different types of <code>File</code></li>
</ul>
</li>
<li><code>genModules</code>:<ul>
<li>Relies on <code>genMain</code>, <code>genInputMod</code>, <code>checkConstClass</code>, <code>genCalcMod</code>, <code>genOutputMod</code>, and <code>genModDef</code></li>
</ul>
</li>
</ul>
</details>

## Summary:
These are the 'root causes' that are stuck not because of GOOL/GProc-related types, but because of `drasil-code`/`drasil-lang` reasons.
- `variable`
	- If the variable is a constant and constant representation is `Const`, then for GOOL it's a `staticVar` and for GProc it's a `constant`.  We might be able to make both use `constant`, I'm not sure.
- `inputVariable`
	- If inputs are bundled, for GOOL it's an object and for GProc it's an error
- `constVariable`
	- If constants are bundled, for GOOL it's an object and for GProc it's an error
- `mkVal`
	- `CodeVarChunk`s have two fields: a `CodeChunk` for the variable and a `Maybe CodeChunk` for an object holding the variable.  GProc can't do objects.
	- `codeType` can also return `Object` because `Actors` are mapped to `Object`, so GOOL uses `convTypeOO` and `GProc` uses `convType`.
- `mkVar`
	- `CodeVarChunk`s have two fields: a `CodeChunk` for the variable and a `Maybe CodeChunk` for an object holding the variable.  GProc can't do objects.
- `genFunc`
	- The `Func` type can either be a function or a constructor.  GProc can't do constructors.
- `convExpr`
	- `CodeExpr` has the `New`, `Message`, and `Field` constructors, each of which has to do with objects, which GProc can't do.
- `fApp`
	- With GOOL, a function may be a private method of the current object, so there's an extra case for a self method call.
- `fAppInOut`
	- With GOOL, a function may be a private method of the current object, so there's an extra case for a self method call.
- `initConsts`
	- If constants are bundled, GOOL declares a new object but GProc throws an error
- `getInputDecl`
	- If inputs are bundled, GOOL declares a new object but GProc throws an error

## Analysis:
There are a few common problems:
- Bundling: GProc can't do bundling, and if it could it would use `struct`s instead of objects.
- `CodeVarChunk`s implicitly support the idea of objects with their `Maybe CodeChunk` field.
- `Space`s implicitly support the idea of objects with their `Actor` value.
- The `Func` type from the Mod AST includes a `CtorDef` value for constructors.
- `CodeExpr` has fields for accessing constructors, methods, and members of a class/object.
- I'm not exactly sure what's going on with `fApp`, but it has a possibility of calling a method on `self`.

Basically, these are all legitimate reasons that not only do the implementations of these functions need to be different, their return types (and those of any functions that use them) need to be different.  I'm not really sure what we can do about that.
- We could try splitting things up a little bit - e.g. split up `CodeVarChunk` into two data types so that one of them can be safely shared.  I'm just not sure how effective that would actually be in most of these cases.  There might be more digging that needs to be done here.
- I'm just thinking off the top of my head, but it seems to me that GOOL was so easy to split up because of its tagless approach, or at least because it uses typeclasses.  Would we be able to get the polymorphism we need by doing something along the same lines here?

**Reply by JacquesCarette** [reply]
Excellent analysis. Clearly too late to do anything about this right now.

Your last point is exactly right: we should be able to use typeclasses to get extra polymorphism to be able to deal with this. The second part of the solution would be to make a number of the above functions more 'semantic' and less code- or feature-centric. Your example of `variable` / `constant` goes in this direction.

Basically, we'd want to make sure we represent "ideas" in Drasil (especially in `drasil-lang` and `drasil-code`) and let the backend decide how to turn that into working code for each language.
